contracts/version-management-design.md

```diff
@@ -0,0 +1,102 @@
+# On-Chain Verifier Upgrade and Deprecation
+
+At base, the RISC Zero verifier contracts are designed to be immutable and stateless.
+This meets the highest standards for trust minimized application on Ethereum, as no action can alter the operation of the verifier.
+
+However, in the event that a security vulnerability is discovered in the verifier contract(s), stateless contracts require the application to respond to maintain security.
+As a countermeasure to this risk, and to allow for upgrades with new features on an opt-in basis, the following version management system is deployed to Ethereum.
+
+## Overview
+
+```mermaid
+---
+title: Smart Contract Relationships
+---
+flowchart LR
+  subgraph IRiscZeroVerifier
+    router["RiscZeroVerifierRouter [managed]"]
+    appRouter["RiscZeroVerifierRouter [app]"]
+    
+    subgraph impls[RiscZeroEmergencyStop]
+      groth16v1["RiscZeroGroth16Verifier [v1]"]
+      aggv1["RiscZeroAggregateVerifier [v1]"]
+      groth16v2["RiscZeroGroth16Verifier [v2]"]
+
+      fflonkv1["RiscZeroFflonkVerifier [v1]"]
+    end
+
+    router --> groth16v1
+    router & appRouter --> groth16v2
+    router --> aggv1
+    router & appRouter --> fflonkv1
+  end
+  timelock[TimelockController]
+  multisig["RISC Zero Multisig"]
+  appAdmin[App Admin]
+  
+  timelock -- admin --> router
```

Comment:
Curiosity: Are the terms admin, proposer, and guardian standard in smart contract development, or are those our terms?

Reply:
Guardian I've seen before in this usage. Proposer is the name used in the `TimelockController` interface. Admin is a term I chose for this.